packages/contracts/contracts/universal/CrossDomainMessenger.sol

```diff
@@ -0,0 +1,244 @@
+    uint256 internal msgNonce;
+
+    /// @notice Address of the CrossDomainMessenger on the other chain.
+    address public otherMessenger;
+
+    /// @notice Mapping of message hash to boolean receipt value.
+    mapping(bytes32 => bool) public receivedMessages;
+
+    /// @notice Blocked system addresses that cannot be called (for security reasons).
+    mapping(address => bool) public blockedSystemAddresses;
+
+    /********************
+     * Public Functions *
+     ********************/
+
+    /**
+     * Pause relaying.
+     */
+    function pause() external onlyOwner {
+        _pause();
+    }
+
+    /**
+     * Retrieves the address of the x-domain message sender. Will throw an error if the sender is
+     * not currently set (equal to the default sender).
+     *
+     * @return Address of the x-domain message sender.
+     */
+    function xDomainMessageSender() external view returns (address) {
+        require(
+            xDomainMsgSender != Lib_DefaultValues.DEFAULT_XDOMAIN_SENDER,
+            "xDomainMessageSender is not set"
+        );
+
+        return xDomainMsgSender;
+    }
+
+    /**
+     * Retrieves the next message nonce. Adds the hash version to the nonce.
+     *
+     * @return Next message nonce with added hash version.
+     */
+    function messageNonce() public view returns (uint256) {
+        return CrossDomainHashing.addVersionToNonce(msgNonce, MESSAGE_VERSION);
+    }
+
+    /**
+     * @param _target Target contract address.
+     * @param _message Message to send to the target.
+     * @param _minGasLimit Gas limit for the provided message.
+     */
+    function sendMessage(
+        address _target,
+        bytes memory _message,
+        uint32 _minGasLimit
+    ) external payable {
+        // TODO: Enforce minimum gas limit.
+
+        _sendMessage(
+            otherMessenger,
+            _minGasLimit, // TODO: Pad this value.
+            msg.value,
+            CrossDomainHashing.getVersionedEncoding(
+                messageNonce(),
+                msg.sender,
+                _target,
+                msg.value,
+                _minGasLimit,
+                _message
+            )
+        );
+
+        emit SentMessage(_target, msg.sender, _message, messageNonce(), _minGasLimit);
+
+        unchecked {
+            ++msgNonce;
+        }
+    }
+
+    function relayMessage(
+        uint256 _nonce,
+        address _sender,
+        address _target,
+        uint256 _value,
+        uint256 _minGasLimit,
+        bytes calldata _message
+    ) external payable nonReentrant whenNotPaused {
+        bytes32 versionedHash = CrossDomainHashing.getVersionedHash(
+            _nonce,
+            _sender,
+            _target,
+            _value,
+            _minGasLimit,
+            _message
+        );
+
+        if (_isSystemMessageSender()) {
+            // Should never happen.
+            require(msg.value == _value, "Mismatched message value.");
```

Comment:
I have concerns about this when we are doing stuff on L2 with this. Specifically the L1 value is set to the L2 mint, and then a user specified (to the deposit tx at least) amount is set to the value field. If these fields do not line up exactly, we have have stuck messages. In particular, we expect one version of the Optimism Portal Deposit function to take eth to pay for L2 gas which would change the amount of value emitted.